include/graybat/Cage.hpp

```diff
@@ -683,6 +692,29 @@ namespace graybat {
     template<typename T_CommunicationPolicy, typename T_GraphPolicy>
     template<typename T>
     auto Cage<T_CommunicationPolicy, T_GraphPolicy>::
+	recv (const Edge &edge, std::future<T>& data)
+	-> void
+	{
+		VAddr srcVAddr = locateVertex(edge.source);
+		std::promise<T> dataPromise;
+		data = dataPromise.get_future();
+
+		// The reason for the std::promise and std::thread over std::async ist the destructor behaviour of std::async
+		// For std::async the destructor is waiting for the completion of the asyncThread, which may lead to a permanent lock
+		// The future from std::promise does not block on the completion of the execution thread.
+		std::thread asyncRecvThread([&](){
+			T result;
+			Event e = comm->asyncRecv(srcVAddr, edge.id, graphContext, result);
+			e.wait();
+			dataPromise.set_value(result);
+		});
+		asyncRecvThread.detach();
+	}
```

Comment:
I understand what you are doing, but maybe you can help me with why you are doing it?

From my point of view, the cage already supports asynchronous communication with ` recv(
        const Edge &edge, T &data, std::vector<Event> &events)`. It might not be the best interface, but its something we can talk about and change. The problem with the lines above is the uncontrolled threading. Each receive call will spawn its own thread, which is not bound to anything. Can we come up with a thread pool like solution (maybe have a look at the io_service of boost::asio) like:
```c++
threadPool.post([&](){
		T result;
		Event e = comm->asyncRecv(srcVAddr, edge.id, graphContext, result);
		e.wait();
		dataPromise.set_value(result);
);
```
Putting this into its own thread policy would be even nicer.